internal/service/alarms/internal/dictionary_definition/dictionary_definition.go

```diff
@@ -0,0 +1,460 @@
+		for _, group := range promRule.Spec.Groups {
+			for _, rule := range group.Rules {
+				// Only alerting rules are of interest (not recording rules)
+				if rule.Alert != "" {
+					rules = append(rules, rule)
+				}
+			}
+		}
+	}
+
+	return rules, nil
+}
+
+// syncDictionaries synchronizes the alarm dictionaries in the database
+func (r *AlarmDictionaryDefinition) syncDictionaries(ctx context.Context, nodeClusterTypes []clusterserver.NodeClusterType) error {
+	slog.Info("Synchronizing alarm dictionaries in the database")
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Minute) // let's try to finish init with a set time.
+	defer cancel()
+
+	// First clean up outdated dictionaries
+	if err := r.deleteOutdatedDictionaries(ctx, nodeClusterTypes); err != nil {
+		return fmt.Errorf("failed to delete outdated dictionaries: %w", err)
+	}
+
+	// Then process all NodeClusterTypes concurrently
+	if err := r.processNodeClusterTypes(ctx, nodeClusterTypes); err != nil {
+		return fmt.Errorf("failed to process resource types: %w", err)
+	}
+
+	slog.Info("Alarm dictionaries and corresponding AlarmDefinitions synchronized")
+	return nil
+}
+
+// deleteOutdatedDictionaries remove any dictionary that may not be available anymore
+func (r *AlarmDictionaryDefinition) deleteOutdatedDictionaries(ctx context.Context, nodeClusterTypes []clusterserver.NodeClusterType) error {
+	// Delete Dictionaries that do not have a corresponding resource type
+	ids := make([]any, 0, len(nodeClusterTypes))
+	for _, nodeClusterType := range nodeClusterTypes {
+		ids = append(ids, nodeClusterType.NodeClusterTypeId)
+	}
+
+	if err := r.AlarmsRepository.DeleteAlarmDictionariesNotIn(ctx, ids); err != nil {
+		return fmt.Errorf("failed to delete alarm dictionaries: %w", err)
+	}
+
+	slog.Info("Outdated alarm dictionaries deleted successfully")
+	return nil
+}
+
+// processNodeClusterTypes process each nodeClusterType in parallel and return early with error if anything fails
+func (r *AlarmDictionaryDefinition) processNodeClusterTypes(ctx context.Context, nodeClusterTypes []clusterserver.NodeClusterType) error {
+	// Validate all rules exist before starting any processing
+	if err := r.validateRulesExist(nodeClusterTypes); err != nil {
+		return err
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(5) // Limit concurrent operations to avoid overwhelming the database
+
+	for _, rt := range nodeClusterTypes {
+		nodeClusterType := rt // Create new variable to avoid data race
+		g.Go(func() error {
+			return r.processNodeClusterType(ctx, nodeClusterType)
+		})
+	}
+	slog.Info("Waiting for all nodeClusterTypes to processed")
+	return g.Wait() //nolint:wrapcheck
+}
+
+// validateRulesExist assert we have rules to work with that will converted to alarm defs
+func (r *AlarmDictionaryDefinition) validateRulesExist(nodeClusterTypes []clusterserver.NodeClusterType) error {
+	var missingRules []uuid.UUID
+	for _, nct := range nodeClusterTypes {
```

Comment:
This is different to what we had before. If for any reason, we fail to collect rules for a resource type we can simply warn and continue collecting the rules for the other resource types. We don't need to stop operation for a single one. Please add it back

Reply:
only warning now